### ISSUE TITLE
Mesh USGS continuous results with WQP discrete results

### ISSUE THREAD

**Issue by cristinamullin** [USEPA/EPATADA#222]
Users would like to retrieve WQP data and OPTIONALLY any additional USGS continuous data that in the specific area of interest and possible also nearby data (within additional buffer zone) or upstream/downstream as well (not just within the bounding box). 

We can bring in USGS continuous data via dataRetrieval. We can also leverage NLDI to query by lat/long (e.g. all sites within 500m). Only problem is that for the continuous data, USGS uses p-codes and not Characteristic Names, so it may be best to wait until after this year (they are working on moving away from p-codes). 

Example code:
[advancedqueries.zip](https://github.com/USEPA/TADA/files/10864478/advancedqueries.zip)

Which is best - NLDI or [nhdplusTools](https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fcran.r-project.org%2Fweb%2Fpackages%2FnhdplusTools%2Findex.html&data=05%7C01%7CMullin.Cristina%40epa.gov%7C58ef2eb6d7944bd5d78008daa2690a87%7C88b378b367484867acf976aacbeca6a7%7C0%7C0%7C638000869350105913%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=fj6XC8edxWVtCWXQVBxre1Bz%2FfmK%2BuGhK5abnFI3Dik%3D&reserved=0) to retrieve catchment geometries in R/ discover nearby data from other data sources?

**Comment by jbousquin** [comment]
Yes on continuous data. For NLDI could you leverage existing R tools, like nhdplusTools in more of a couple short demo?

**Comment by cristinamullin** [comment]
Note from Shelly: Consider how the NLDI use case would be done with WATERS. NLDI and WATERS both use a manual process to index WQP location to NHDPlus reach codes.  NLDI is stored at USGS and WATERS is stored at EPA. NLDI and WATERS both index WQP locations to NHDPlus V2. WQP indexed locations to the WBD (HUC12s) not NHD or NHDPlus. If we wanted to bring in the USGS continuous data we may want to consider writing a TADA R function that leverages WATERS APIs and USGS continuous data APIs rather than leveraging dataRetreival (which includes the NLDI integration: https://github.com/DOI-USGS/dataRetrieval/blob/main/R/findNLDI.R). Needs more discussion!

**Comment by cristinamullin** [comment]
[OP_Mapping_forEPA.txt](https://github.com/USEPA/TADA/files/12578317/OP_Mapping_forEPA.txt)
Note that USGS migration from p-codes to observed properties may fix that compatibility issue. Not sure if this migration to observed properties will cover NWIS cont data or only their discrete WQ data.

**Comment by kathryn-willi** [comment]
1) For new TADA_listNWIS() and TADA_getNWIS() functions, explore editing the search 'parameters' argument to instead be observed properties. 
2) Modify the returned results of TADA_listNWIS() and TADA_getNWIS() to be compatible with WQX continuous data call results.

**Comment by cristinamullin** [comment]
There are some global variable notes. We can address these by removing the variable using **rm()** once the variable is no longer needed.

```r
   TADA_GetATTAINS: no visible binding for global variable 'count'
   TADA_GetATTAINS : find_distances: no visible binding for global
     variable 'TADA.DistanceAway.Meters'
   TADA_GetATTAINS: no visible binding for global variable
     'TADA.DistanceAway.Meters'
   TADA_GetATTAINS: no visible global function definition for
     'st_drop_geometry'
   TADA_getNWIS: no visible binding for global variable 'dec_long_va'
   TADA_getNWIS: no visible binding for global variable 'dec_lat_va'
   TADA_getNWIS: no visible binding for global variable 'site_no'
   TADA_getNWIS: no visible binding for global variable 'agency_cd'
   TADA_getNWIS: no visible binding for global variable 'Date'
   TADA_getNWIS: no visible binding for global variable 'NWIS.parameter'
   TADA_getNWIS: no visible binding for global variable 'NWIS.value'
   TADA_getNWIS: no visible binding for global variable 'NWIS.status'
   TADA_listNWIS : pcodes: no visible binding for global variable
     'parameter_code'
   TADA_listNWIS: no visible binding for global variable 'dec_long_va'
   TADA_listNWIS: no visible binding for global variable 'dec_lat_va'
   TADA_listNWIS: no visible binding for global variable 'site_no'
   TADA_listNWIS: no visible binding for global variable 'station_nm'
   TADA_listNWIS: no visible binding for global variable 'site_type'
   TADA_listNWIS: no visible binding for global variable 'site_tp_cd'
   TADA_listNWIS: no visible binding for global variable 'data_type'
   TADA_listNWIS: no visible binding for global variable 'data_type_cd'
   TADA_listNWIS: no visible binding for global variable
     'parameter_name_description'
   TADA_listNWIS: no visible binding for global variable 'parm_cd'
   TADA_listNWIS: no visible binding for global variable 'count_nu'
   TADA_listNWIS: no visible binding for global variable 'begin_date'
   TADA_listNWIS: no visible binding for global variable 'end_date'
   fetchATTAINS : perform_iterative_clustering : bbox_area: no visible
     binding for global variable 'cluster'
   fetchATTAINS : perform_iterative_clustering : split_clusters_by_area:
     no visible binding for global variable 'cluster'
   fetchATTAINS: no visible binding for global variable 'cluster'
   Undefined global functions or variables:
     Date NWIS.parameter NWIS.status NWIS.value TADA.DistanceAway.Meters
     agency_cd begin_date cluster count count_nu data_type data_type_cd
     dec_lat_va dec_long_va end_date parameter_code
     parameter_name_description parm_cd site_no site_tp_cd site_type
     st_drop_geometry station_nm
```

**Comment by cristinamullin** [comment]
Note from Laura D (USGS): 

I have an article here on how to join by the closest date/time for a discrete WQP result and USGS continuous result:
[https://doi-usgs.github.io/dataRetrieval/articles/join_by_closest.html](https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdoi-usgs.github.io%2FdataRetrieval%2Farticles%2Fjoin_by_closest.html&data=05%7C02%7CMullin.Cristina%40epa.gov%7C228fae99a9aa447cc38d08ddb8e8e98a%7C88b378b367484867acf976aacbeca6a7%7C0%7C0%7C638870032683714470%7CUnknown%7CTWFpbGZsb3d8eyJFbXB0eU1hcGkiOnRydWUsIlYiOiIwLjAuMDAwMCIsIlAiOiJXaW4zMiIsIkFOIjoiTWFpbCIsIldUIjoyfQ%3D%3D%7C0%7C%7C%7C&sdata=DdTs1FY0qkFAzZOeZdQmzy75q8p4IO6xyvc7jGzQiaU%3D&reserved=0)
Note, all the "NWIS" functions are going to be going away soon, so I wouldn't start any big work on the "readNWISuv" function just yet. The replacement will probably be "read_waterdata_continuous" and should come out...I think this fall?

I have an internal Shiny app that does some nice hand-holding on joining WQP with continuous data. I have the reviews to flip it to public, just need the time to respond. If I get that done, I'll pass it on to you.
